File: 0_array/question/threeSum/best.cpp
#include <iostream>
#include <vector>
#include <set>
 
using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
             sort(nums.begin(), nums.end());

          vector<vector<int>> res;
          for(int i = 0 ; i < nums.size() ; i++){
             //!m2::  Skip duplicate values for the first number
            if (i > 0 && nums[i] == nums[i - 1]) continue;

            int j = nums.size() - 1 ;
            int k = i + 1 ;
            while (k < j) {
                int target = nums[i] + nums[j] + nums[k];

                if(target > 0 ){
                    j--;
                }else if(target < 0){
                    k++;
                }else{
                            res.push_back({nums[i], nums[k], nums[j]});
                       while(k < nums.size() - 1 && nums[k] == nums[k+1]){
                        k++;
                      }
                     while(j > 0 && nums[j] == nums[j-1]){
                        j--;
                      }
 
                    k++;
                    j--;
                }
                }
          }
     

        return res;
    }
};

      

File: 0_array/question/threeSum/best.js
var threeSum = function (nums) {
    let res = []
   nums = nums.sort((a, b) => a - b)
   for (let i = 0; i < nums.length; i++) {
     if (i > 0 && i < nums.length && nums[i] === nums[i - 1]) continue
 
     let j = i + 1
     let k = nums.length - 1
     //^ = not k less than j
     while (j < k) {
       let sum = nums[i] + nums[j] + nums[k]
       if (sum < 0) {
         j++
       } else if (sum > 0) {
         k--
       } else if (sum === 0) {
         res.push([nums[i], nums[j], nums[k]]) //! m1: place below j++,k--
 
         j++
         k--
         //!m2: j - 1 missed
         //!m3: j < k
         while (j < nums.length && nums[j] === nums[j - 1]) j++
         while (k > 0 && nums[k] === nums[k + 1]) k--
       }
     }
   }
   return res
 }